pkg/storegateway/indexheader/index/postings.go

```diff
@@ -473,22 +481,25 @@ func postingOffsets[T any](t *PostingOffsetTableV2, name string, prefix string,
 		return nil, nil
 	}
 
-	offsetIdx := 0
+	offsetsStart, offsetsEnd := 0, len(e.offsets)
 	if prefix != "" {
-		offsetIdx, ok = e.prefixOffset(prefix)
+		offsetsStart, offsetsEnd, ok = e.prefixOffsets(prefix)
 		if !ok {
 			return nil, nil
 		}
 	}
-	result := make([]T, 0, (len(e.offsets)-offsetIdx)*t.postingOffsetsInMemSampling)
+	result := make([]T, 0, (offsetsEnd-offsetsStart)*t.postingOffsetsInMemSampling)
 
 	// Don't Crc32 the entire postings offset table, this is very slow
 	// so hope any issues were caught at startup.
 	d := t.factory.NewDecbufAtUnchecked(t.tableOffset)
 	defer runutil.CloseWithErrCapture(&err, &d, "get label values")
 
-	d.ResetAt(e.offsets[offsetIdx].tableOff)
+	d.ResetAt(e.offsets[offsetsStart].tableOff)
 	lastVal := e.offsets[len(e.offsets)-1].value
+	if offsetsEnd < len(e.offsets) {
+		lastVal = e.offsets[offsetsEnd].value
+	}
```

Comment:
Question: to my understanding the `offsetsEnd` returned by `e.prefixOffsets()` is the first index not matching the prefix, but here we want the last value matching the prefix. Shouldn't we change it as follow?

```suggestion
	lastVal := e.offsets[offsetsEnd-1].value
```

Reply:
If I'm correct, I would also add a comment to explain it, cause it may not be obvious just reading the code.